### PR TITLE
Support subcomponent

### PR DESCRIPTION
**Pull request by hl4**
This may be big. It includes serveral related issue.
* improved the code to remove prefix './' from component root path by using path::is_dot_filename()
* fixed sub-component name. for example, path 'module1/submodule1/src', it were make libaray in name libmodule1_submodule1.a in gcc style. 
* using a separated field 'name' for component name. this may be improved more.
* add unit test for above code. using boost.test library. those has not been included in cmake/Makefile yet. only evoke invoke those. yes!

Shall we leave android toolset code alone, since there is a separate branch to incubate it? It's unstructed and different from others.

### CODE REVIEW

File: project/test/main.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// mmap
+#define BOOST_TEST_MODULE Evoke_project
+#include "Component.h"
+
+#include <boost/test/included/unit_test.hpp>
+
+namespace btt = boost::test_tools;
+
+BOOST_AUTO_TEST_CASE(without_last_flash)
```
Comment (dascandy) [inline project/test/main.cpp]: Flash? Did you mean slash?

File: toolsets/src/Toolset.cpp [hunk 2]
```diff
@@ -9,11 +9,7 @@
 
 std::string Toolset::getNameFor(const Component &component)
 {
-    if(component.root.string() != ".")
-    {
-        return as_dotted(component.root.generic_string());
-    }
-    return filesystem::weakly_canonical(component.root).filename().string();
+    return component.GetName();
```
Comment (dascandy) [inline toolsets/src/Toolset.cpp]: You may be able to remove this whole function now.

File: toolsets/src/gcc.cpp [hunk 3]
```diff
@@ -49,7 +49,11 @@ std::string GccToolset::getLibNameFor(const Component &component)
 
 std::string GccToolset::getExeNameFor(const Component &component)
 {
+#if defined(_WIN32)
+    return getNameFor(component) + ".exe";
```
Comment (dascandy) [inline toolsets/src/gcc.cpp]: Can you add an Appveyor config for this? I believe this is MingW, right?

**Comment by dascandy** [comment]
This looks really good. Seeing you add Boost.Test test cases without any build config, working automatically in Evoke, while I hadn't tried Boost.Test yet makes me very happy. Things should *just work* and this seems to confirm this is a good direction.

**Comment by dascandy** [comment]
Oh, yes, please do ignore Android for now. As you said, it's the odd one out and its fate will most likely be that it'll be removed from Evoke.

File: toolsets/src/gcc.cpp (hunk 3 above)
Comment (hl4) [reply toolsets/src/gcc.cpp]: see pr #56

**Pull request merged**